crates/axvm/src/vm.rs

```diff
@@ -177,48 +195,75 @@ impl<H: AxVMHal> AxVM<H> {
         Ok(image_load_hva)
     }
 
+    pub fn running(&self) -> bool {
+        self.running.load(Ordering::Relaxed)
+    }
+
     pub fn boot(&self) -> AxResult {
         if !has_hardware_support() {
             ax_err!(Unsupported, "Hardware does not support virtualization")
+        } else if self.running() {
+            ax_err!(BadState, format!("VM[{}] is running", self.id()))
         } else {
-            self.run_vcpu(0)
+            info!("Booting VM[{}]", self.id());
+            self.running.store(true, Ordering::Relaxed);
+            Ok(())
         }
     }
 
     pub fn get_device_list(&self) -> &mut AxArchDeviceList<H> {
         unsafe { &mut *self.inner_const.device_list.get() }
     }
 
-    pub fn run_vcpu(&self, vcpu_id: usize) -> AxResult {
+    pub fn run_vcpu(&self, vcpu_id: usize) -> AxResult<VCpuStatus> {
         let vcpu = self
             .vcpu(vcpu_id)
             .ok_or_else(|| ax_err_type!(InvalidInput, "Invalid vcpu_id"))?;
         vcpu.bind()?;
+
+        let mut vcpu_status = VCpuStatus::Running;
+
         loop {
-            // todo: device access
             let exit_reason = vcpu.run()?;
 
+            trace!("{exit_reason:#x?}");
             match exit_reason {
-                axvcpu::AxArchVCpuExitReason::MmioRead { addr, width: _ } => {
-                    debug!("EPT from addr {:#x}", addr);
-
-                    let paddr = self
-                        .inner_mut
-                        .address_space
-                        .lock()
-                        .translate(VirtAddr::from(addr));
-
-                    debug!("EPT mapped to {:#x?}", paddr);
-
-                    break;
+                AxVCpuExitReason::Hypercall { nr, args } => {
+                    debug!("Hypercall [{}] args {:x?}", nr, args);
+                }
+                AxVCpuExitReason::Nothing => {}
+                AxVCpuExitReason::Halt => {
+                    vcpu_status = VCpuStatus::Yield;
+                }
+                AxVCpuExitReason::FailEntry {
+                    hardware_entry_failure_reason: _,
+                } => {
+                    vcpu_status = VCpuStatus::Exit;
                 }
                 _ => {
                     let device_list = self.get_device_list();
                     device_list.vmexit_handler(vcpu.get_arch_vcpu(), exit_reason)?;
                 }
             }
+            if vcpu_status != VCpuStatus::Running {
+                break;
+            }
         }
         vcpu.unbind()?;
-        panic!("VCpu [{}] halt", vcpu.id())
+        Ok(vcpu_status)
     }
 }
+
+numeric_enum_macro::numeric_enum! {
+#[repr(u32)]
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[allow(non_camel_case_types)]
+/// VCpu status after a VM Exit.
+/// Todo
+pub enum VCpuStatus {
```

Comment:
Can we merge this enum with `VCpuState`?

Reply:
It's confusing

Reply:
We need to return a status for current execution flow of axtask (the "big loop{...}") to determine whether the vCPU running on the current task needs to be scheduled.

Reply:
Is it necessary here?
The guest should trigger some interrupts to change the task's running state, right? For example, a wfi interrupt trap or psci?

Reply:
Yep, the logic has been improved and will keep improving

Reply:
Maybe it needs a better name to avoid confusion.

Reply:
Currently I just combined it with `AxVCpuExitReason`.

Basic VM-Exit handling sequence is:
arch_related VM-Exit(arch bulitin handler) -> axvm VM-Exit (emulated devices) > VM-Exit handled in axtask loop{}